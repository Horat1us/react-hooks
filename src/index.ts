export * from "./use-open-state";
export * from "./use-window-width";
export * from "./use-interval";
export * from "./use-timeout";
export * from "./use-submit-callback";
export * from "./use-input-state";
