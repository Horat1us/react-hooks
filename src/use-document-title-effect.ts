import * as React from "react";

export function useDocumentTitleEffect(title: string): void {
    React.useEffect(() => {
        document.title = title;
    }, [title]);
}
