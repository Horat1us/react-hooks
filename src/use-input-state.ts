import * as React from "react";

export type InputFilter = (rawInput: string) => string;
export type InputType = HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement;
export type InputStateHook<I extends InputType = InputType, S extends React.HTMLProps<I>["value"] = string> =
    (initialState?: S, inputFilter?: InputFilter) => [S, React.Dispatch<React.ChangeEvent<I>>, React.Dispatch<React.SetStateAction<S>>];

export const useInputState: InputStateHook = (initialState = "", inputFilter) => {
    const [state, setState] = React.useState(initialState);
    const handleChange = React.useCallback(
        ({ target: { value } }) => setState(inputFilter ? inputFilter(value) : value),
        [setState, inputFilter]
    );
    return [state, handleChange, setState];
};
