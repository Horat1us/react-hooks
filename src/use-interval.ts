import * as React from "react";

export const useInterval = (handler?: () => void, timeout?: number): void =>{
    React.useEffect(() => {
        if (!handler) {
            return;
        }
        const id = setInterval(handler, timeout);
        return () => clearInterval(id);
    }, [handler, timeout]);
};
