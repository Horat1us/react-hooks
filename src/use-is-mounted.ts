import * as React from "react";

export const useIsMounted = () => {
    var isMounted = React.useRef(true);
    React.useEffect(() => {
        return () => {
            isMounted.current = false;
        };
    }, []);
    return isMounted.current;
}
