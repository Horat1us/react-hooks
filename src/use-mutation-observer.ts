import * as React from "react";

export function useMutationObserver<E extends HTMLElement = HTMLElement>(
    cb: (...args: Parameters<MutationCallback>) => ReturnType<React.EffectCallback>
): [E | null, (element: E | null) => void] {
    const [ref, setRef] = React.useState<E | null>(null);

    React.useEffect(() => {
        if (!ref) {
            return;
        }
        let cancelCallback = () => observer.disconnect();
        const observer = new MutationObserver((...args: Parameters<MutationCallback>) => {
            const observerCancelCallback = cb(...args);
            if (observerCancelCallback) {
                const mutationCancelCallback = cancelCallback;
                cancelCallback = () => {
                    mutationCancelCallback();
                    observerCancelCallback();
                };
            }
        });
        observer.observe(ref, {childList: true, subtree: true});
        return cancelCallback;
    }, [ref, cb]);

    return [ref, setRef];
}
