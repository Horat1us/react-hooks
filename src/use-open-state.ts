import * as React from "react";

export const openStateReducer = (prevState: boolean): boolean => !prevState;
export const useOpenState = (initialState: boolean = false): [boolean, () => void] => {
    return React.useReducer<boolean, []>(
        openStateReducer,
        initialState
    ) as [boolean, () => void];
};
