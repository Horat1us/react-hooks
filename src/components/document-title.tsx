import * as React from "react";

import {useDocumentTitleEffect} from "../use-document-title-effect";
import {ForwardRefRenderFunction} from "react";

export type DocumentTitleProps = React.HTMLProps<HTMLTitleElement> & {
    children: string;
}

export const DocumentTitle = React.memo(React.forwardRef<HTMLTitleElement, DocumentTitleProps>(
    (props, ref) => <title {...props} ref={ref}/>
));
DocumentTitle.displayName = "RH.DocumentTitle";
