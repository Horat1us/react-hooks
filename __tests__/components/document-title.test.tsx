import * as React from "react";
import * as ReactDOM from "react-dom";
import {act} from "react-dom/test-utils";
import {DocumentTitle} from "../../src/components";

test("<DocumentTitle />", () => {
    const newTitle = `<DocumentTitle> test case`;
    act(() => {
        ReactDOM.render(<DocumentTitle data-x="s">{newTitle}</DocumentTitle>, document.head);
    });
    expect(document.title).toEqual(newTitle);
    expect(document.head).toMatchSnapshot();
});


