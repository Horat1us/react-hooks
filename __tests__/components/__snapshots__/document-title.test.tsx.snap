// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DocumentTitle /> 1`] = `
<head>
  <title
    data-x="s"
  >
    &lt;DocumentTitle&gt; test case
  </title>
</head>
`;
